Libraries/LibWeb/HTML/StructuredSerialize.cpp

```diff
@@ -1156,8 +1170,32 @@ WebIDL::ExceptionOr<SerializedTransferRecord> structured_serialize_with_transfer
         // IMPLEMENTATION DEFINED: We just create a data holder here, our memory holds indices into the SerializationRecord
         TransferDataHolder data_holder;
 
-        // FIXME 4. If transferable has an [[ArrayBufferData]] internal slot, then:
-        if (false) {
+        // 4. If transferable has an [[ArrayBufferData]] internal slot, then:
+        if (is_array_buffer) {
+            // 1. If transferable has an [[ArrayBufferMaxByteLength]] internal slot, then:
+            auto& array_buffer = dynamic_cast<JS::ArrayBuffer&>(*transferable);
+            if (!array_buffer.is_fixed_length()) {
+                // 1. Set dataHolder.[[Type]] to "ResizableArrayBuffer".
+                // 2. Set dataHolder.[[ArrayBufferData]] to transferable.[[ArrayBufferData]].
+                // 3. Set dataHolder.[[ArrayBufferByteLength]] to transferable.[[ArrayBufferByteLength]].
+                // 4. Set dataHolder.[[ArrayBufferMaxByteLength]] to transferable.[[ArrayBufferMaxByteLength]].
+                serialize_enum<TransferType>(data_holder.data, TransferType::ResizableArrayBuffer);
+                MUST(serialize_bytes(vm, data_holder.data, array_buffer.buffer().bytes())); // serializes both byte length and bytes
+                serialize_primitive_type<size_t>(data_holder.data, array_buffer.max_byte_length());
+            }
+
+            // 2. Otherwise:
+            else {
+                // 1. Set dataHolder.[[Type]] to "ArrayBuffer".
+                // 2. Set dataHolder.[[ArrayBufferData]] to transferable.[[ArrayBufferData]].
+                // 3. Set dataHolder.[[ArrayBufferByteLength]] to transferable.[[ArrayBufferByteLength]].
+                serialize_enum<TransferType>(data_holder.data, TransferType::ArrayBuffer);
+                MUST(serialize_bytes(vm, data_holder.data, array_buffer.buffer().bytes())); // serializes both byte length and bytes
+            }
+
+            // 3. Perform ? DetachArrayBuffer(transferable).
+            // NOTE: Specifications can use the [[ArrayBufferDetachKey]] internal slot to prevent ArrayBuffers from being detached. This is used in WebAssembly JavaScript Interface, for example. See: https://html.spec.whatwg.org/multipage/references.html#refsWASMJS
+            MUST(JS::detach_array_buffer(vm, array_buffer));
```

Comment:
The ? In the call here means we should be using TRY() instead of MUST. unless you have a reason that this call will never fail? If that's the case, a comment and a possible spec bug are in order to change it to `! DetachArrayBuffer`